presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkMutableRowPageInput.java

```diff
@@ -108,23 +76,34 @@ else if (row.getArray() != null) {
                     }
                     rowCount++;
                 }
-                else {
+                if (!iterator.hasNext()) {
                     currentIteratorIndex++;
                 }
+                else {
+                    break;
```

Comment:
Note this means when one remote source is consumed, we will return (instead of try to fill `TARGET_SIZE` or `MAX_ROWS_PER_PAGE`). Curious if there is any reason for this? :) 

(the code has the same semantic than before)

Reply:
I see,  it will be easier to understand using 
`
if (output.size() > TARGET_SIZE || rowCount > MAX_ROWS_PER_PAGE) {
break;}
else {
  // stass update
}
`

Reply:
Explained it here:

https://github.com/prestodb/presto/pull/14938#discussion_r465873803

> I see, it will be easier to understand using

If feels like this is about the same. In the first place we check if the iterator is done, so we need to start a new one. In the second we check if the buffer is full. I would prefer to keep it the current way.